### PR TITLE
Backend interface signer service will need access to

### PR DESCRIPTION
**Pull request by Gozala**

### CODE REVIEW

File: packages/query/src/provider.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
+import type { Link, DID } from '@ipld/dag-ucan'
```
Comment (Gozala) [inline packages/query/src/provider.ts]:
```suggestion
```

File: packages/query/src/provider.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    cid: Link<T>
+  ): Result<void, ProofNotFoundError | RevokedError | InvalidProofError>
+
+  /**
+   * Saves valid proof and associated token. Calling `get` after this should return `T`
+   * unless it has been rovked. It may error with `RevokedError` if it had
+   * been revoked concurrently by other actor.
+   *
+   */
+  validate(cid: Link<T>, token: T): Result<undefined, RevokedError>
```
Comment (Gozala) [inline packages/query/src/provider.ts:30]:
```suggestion
  storeValid(cid: Link<T>, token: T): Result<undefined, RevokedError>
```

File: packages/query/src/provider.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+  /**
+   * Saves invalid proof so that we can avoid validating it again.
+   */
+  invalidate(
```
Comment (Gozala) [inline packages/query/src/provider.ts:35]:
```suggestion
  storeInvalid(
```

File: packages/query/src/provider.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+   */
+  get(
+    cid: Link<T>
+  ): Result<T, ProofNotFoundError | RevokedError | InvalidProofError>
+
+  /**
+   * Checks status of the given proof by it's CID. Just like get execept when
+   * actual value is irrelevant.
+   */
+  status(
```
Comment (Gozala) [inline packages/query/src/provider.ts]:
```suggestion
  getStatus(
```

File: packages/query/src/provider.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  proofs: ProofsProvider<Token>
+  store: StoreProvider<DID, Link, Token>
+}
+
+export interface ProofsProvider<T> {
+  /**
+   * Service MAY need to lookup a proof by CID that was previously stored. In
+   * such case it will call this to do a lookup.
+   */
+  get(
```
Comment (Gozala) [inline packages/query/src/provider.ts]:
```suggestion
  getToken(
```

File: packages/query/src/provider.ts [hunk 6]
```diff
@@ -0,0 +1,92 @@
+
+export interface Context<DID, Link, Token> {
+  proofs: ProofsProvider<Token>
+  store: StoreProvider<DID, Link, Token>
+}
+
+export interface ProofsProvider<T> {
```
Comment (Gozala) [inline packages/query/src/provider.ts]:
```suggestion
export interface TokenProvider<T> {
```

File: packages/query/src/provider.ts [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    ucan: Proof
+  ): Result<undefined, UnauthorizedDIDError | DoesNotHasError>
+}
+
+export interface AddStatus {
+  /**
+   * Should be `ok` if we already have car and we don't need to perform upload.
+   * Otherwise should be `pending`.
+   */
+  status: 'ok' | 'pending'
```
Comment (Gozala) [inline packages/query/src/provider.ts]:
```suggestion
  status: 'in-s3' | 'not-in-s3'
```

File: packages/query/src/provider.ts [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+   * Saves invalid proof so that we can avoid validating it again.
+   */
+  invalidate(
+    cid: Link<T>,
+    token: T,
+    reason: InvalidProofError
+  ): Result<undefined, ProofNotFoundError>
+}
+
+export interface StoreProvider<DID, Link, Proof> {
```
Comment (Gozala) [inline packages/query/src/provider.ts]:
```suggestion
export interface DIDLinkAssociationProvider<DID, Link, Proof> {
```

File: packages/query/src/provider.ts [hunk 9]
```diff
@@ -0,0 +1,92 @@
+
+export interface Context<DID, Link, Token> {
+  proofs: ProofsProvider<Token>
+  store: StoreProvider<DID, Link, Token>
```
Comment (Gozala) [inline packages/query/src/provider.ts]:
```suggestion
  store: DIDLinkAssociationProvider<DID, Link, Token>
```

**Review by the-simian — APPROVED** [review]
Just annotating here that we went over this line by line (Ben, Gozala and I) and we feel really good about this approach, so thumbs up from me

**Pull request closed without merging**